flow/project.py

```diff
@@ -265,8 +277,135 @@ def make_bundles(operations, size=None):
+        Information on how to aggregate jobs.
+    :type grouper:
+        callable
+    :param sort:
+        Sort jobs by a statepoint parameter.
+    :type sort:
+        str or NoneType
+    :param reverse:
+        States if the jobs are to be sorted in reverse order.
+    :type reverse:
+        bool
+    """
+
+    def __init__(self, grouper=None, sort=None, reverse=False):
+        if grouper is None:
+            def grouper(jobs):
+                yield jobs
+
+        def key_sort(job):
+            try:
+                return job.sp[sort]
+            except KeyError:
+                raise KeyError("The key '{}' was not found in statepoint "
+                               "parameters of the job {}.".format(sort, job))
+
+        if not callable(grouper):
+            raise TypeError("Expected callable grouper function, got {}".format(type(grouper)))
+
+        if sort is not None and not isinstance(sort, str):
+            raise TypeError("Expected string sort parameter, got {}".format(type(sort)))
+
+        if not isinstance(reverse, bool):
+            raise TypeError("Expected bool reverse parameter got {}".format(type(reverse)))
+
+        self._grouper = grouper
+        self._sort = None if sort is None else functools.partial(sorted,
+                                                                 key=key_sort,
+                                                                 reverse=reverse)
+
+    @classmethod
+    def groupsof(cls, num=1, sort=None, reverse=False):
+        # copied from: https://docs.python.org/3/library/itertools.html#itertools.zip_longest
+        def grouper(jobs):
+            args = [iter(jobs)] * num
+            return zip_longest(*args)
+
+        return cls(grouper, sort, reverse)
+
+    @classmethod
+    def groupby(cls, key=None, default=None, sort=None, reverse=False):
+        if isinstance(key, str):
+            if default is None:
+                def keyfunction(job):
+                    return job.sp[key]
+            else:
+                def keyfunction(job):
+                    return job.sp.get(key, default)
+
+        elif isinstance(key, Iterable):
+            key = list(key)
+            if default is None:
+                def keyfunction(job):
+                    return [job.sp[k] for k in key]
+            else:
+                def keyfunction(job):
+                    return [job.sp.get(k, default) for k in key]
+
+        elif key is None:
+            # Must return a type that can be ordered with <, >
+            def keyfunction(job):
+                return str(job)
+
+        else:
+            keyfunction = key
+
+        def grouper(jobs):
+            for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
+                yield group
+
+        return cls(grouper, sort, reverse)
+
+    def __call__(self, func):
+        setattr(func, '_flow_aggregate', (self._grouper, self._sort))
+        return func
+
+
+class select(object):
```

Comment:
Unless we are offering more options, this does not need to be a class like `aggregate`.